### PR TITLE
LibWeb+LibGfx: A first pass at supporting (linear) gradients!

### PR DESCRIPTION
**Pull request by MacDue**
Was bored, needed a new yak... Ran out of `border-radius` stuff :cry: 

image

This PR implements parsing for `linear-gradient()`, along with the the necessary plumbing, and some very basic painting.

The parsing is should work for everything defined in the W3 spec https://drafts.csswg.org/css-images/#linear-gradients, though there seems to be some issues there as multi-position stop points are not defined (though are demonstrated on MDN).
Though we can't paint all but the simplest of gradients yet, so this is not a huge issue yet.

P.S. The CSS parsing and style value side of LibWeb is slightly unfamiliar territory for me, so might not be optimal :^).

### CODE REVIEW

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Very cool! :eyes: Had a little time to look at the first commit, will look at the rest later.

**Comment by MacDue** [comment]
Rebased for the StringView changes from #14555

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 1]
```diff
@@ -2355,6 +2356,184 @@ Optional<AK::URL> Parser::parse_url_function(ComponentValue const& component_val
...
+        if (!tokens.next_token().is(Token::Type::Comma))
+            return {};
+        auto element_type = parse_color_stop_list_element(list_element);
+        if (element_type == ElementType::ColorHint) {
+            // <linear-color-hint>, <linear-color-stop>
+            tokens.skip_whitespace();
+            if (!tokens.next_token().is(Token::Type::Comma))
+                return {};
+            if (parse_color_stop_list_element(list_element) != ElementType::ColorStop)
+                return {};
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2525]: You throw away the ColorHint element here. That's fine for the first implementation to keep things simple, but please leave a `// FIXME` about it for clarity. :^)

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/GenerateCSSPropertyID.cpp [hunk 2]
```diff
@@ -492,6 +492,11 @@ bool property_accepts_value(PropertyID property_id, StyleValue& style_value)
                             property_generator.append(R"~~~(
         if (style_value.is_image())
             return true;
+)~~~");
+                        } else if (type_name == "linear-gradient") {
```
Comment (AtkinsSJ) [inline Meta/Lagom/Tools/CodeGenerators/LibWeb/GenerateCSSPropertyID.cpp:496]: Is this needed? I haven't looked into gradients that much, but from what I can tell a gradient is part of the `<image>` type so should be valid everywhere that images are, without special handling.

Instead, I think the `image` branch above should just check for `style_value.is_linear_gradient()` too. Do correct me if I'm wrong though. :^)

File: Base/res/html/misc/welcome.html [hunk 3]
```diff
@@ -105,6 +105,7 @@ <h2>CSS</h2>
             <li><a href="hover.html">:hover</a></li>
             <li><a href="focus.html">:focus</a></li>
             <li><h3>Properties</h3></li>
+            <li><a href="gradients.html">Gradients</a></li>
```
Comment (AtkinsSJ) [inline Base/res/html/misc/welcome.html:108]: I would combine this commit with the previous one.

File: Userland/Libraries/LibWeb/Painting/BackgroundPainting.cpp [hunk 4]
```diff
@@ -93,6 +133,16 @@ void paint_background(PaintContext& context, Layout::NodeWithStyleAndBoxModelMet
...
 
+        if (layer.background_image->is_linear_gradient()) {
+            // FIXME: Paint non-orthogonal gradients, gradients with > 2 color stops, gradients with custom stop positions ...
+            auto& linear_gradient = layer.background_image->as_linear_gradient();
+            auto gfx_gradient = linear_gradient_to_gfx_gradient(linear_gradient, border_box.rect);
+            if (gfx_gradient.has_value()) {
+                painter.fill_rect_with_gradient(gfx_gradient->orientation, border_box.rect.to_rounded<int>(), gfx_gradient->color_a, gfx_gradient->color_b);
+            }
+            continue;
+        }
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/Painting/BackgroundPainting.cpp:144]: Gradient backgrounds follow the same positioning, clipping and sizing rules as regular images, so this needs to happen after all those calculations. Which is quite a bit of extra work, sorry!

(As always, you can just put a `// FIXME:` on this if you don't want to deal with it right away.)

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
I have a few changes, but this is good stuff. :^)

One issue I'm not sure about, which doesn't have to be fixed here, is how to deal with gradients being allowed anywhere an image is. As written, you'd have to manually enable gradients for each property with specific painting code for them, which feels awkward. Maybe ImageStyleValue and LinearGradientStyleValue can both implement some interface for getting their inherent dimensions and painting them into a rectangle.

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp (hunk 1 above)
Comment (MacDue) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2525]: The color hint is not being thrown away here, `parse_color_stop_list_element()` is (somewhat awkwardly) adding that information to the `list_element` which it takes by reference.

 So If the type is `ColorHint` the `Optional` hint has been set on the `list_element`, then at the second call to `parse_color_stop_list_element()` fills in the color stop in the same list element (a single element can include both a transition hint and a color stop).

Comment (AtkinsSJ) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2525]: Ohhh, you're right, my bad!

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/GenerateCSSPropertyID.cpp (hunk 2 above)
Comment (MacDue) [reply Meta/Lagom/Tools/CodeGenerators/LibWeb/GenerateCSSPropertyID.cpp:496]: I think you're right :)

File: Userland/Libraries/LibWeb/Painting/BackgroundPainting.cpp (hunk 4 above)
Comment (MacDue) [reply Userland/Libraries/LibWeb/Painting/BackgroundPainting.cpp]: The clipping already works, the rest is a +1 :bison: :)

**Comment by MacDue** [comment]
Done the easy changes and added some `FIXMEs` for the rest. I think trying to make gradients and images share a common API for painting would be good, but I think that should be tackled separately since I think it'll be a bit tricky :^)

**Comment by Lubrsi** [comment]
It seems we must also support the `-webkit` prefix for linear-gradient. For example, the GitHub landing page only uses the `-webkit` prefixed version with no fallback:
Screenshot from 2022-07-16 14-51-56

**Comment by MacDue** [comment]
The parser should recognise the `-webkit` variant now too :^)

**Comment by AtkinsSJ** [comment]
LGTM! (Look, Gradients; Their Majesty!)

**Pull request merged**